src/dune_rules/artifact_substitution.ml

```diff
@@ -539,3 +539,69 @@ let copy_file ~conf ?chmod ~src ~dst () =
       Io.close_both (ic, oc);
       Fiber.return ())
     (fun () -> copy ~conf ~input_file:src ~input:(input ic) ~output:(output oc))
+
+let test ~input =
+  let rec loop scanner_state ~pos ~end_of_data =
+    let scanner_state = Scanner.run scanner_state ~buf ~pos ~end_of_data in
+    let placeholder_start =
+      match scanner_state with
+      | Scan0 -> end_of_data
+      | Scan1 -> end_of_data - 1
+      | Scan2 -> end_of_data - 2
+      | Scan_prefix placeholder_start
+      | Scan_length (placeholder_start, _)
+      | Scan_placeholder (placeholder_start, _) ->
+        placeholder_start
+    in
+    let leftover = end_of_data - placeholder_start in
+    match scanner_state with
+    | Scan_placeholder (placeholder_start, len) when len <= leftover -> (
+      let placeholder = Bytes.sub_string buf ~pos:placeholder_start ~len in
+      match decode placeholder with
+      | Some _ -> Fiber.return true
+      | None ->
+        (* Restart just after [prefix] since we know for sure that a placeholder
+           cannot start before that. *)
+        loop Scan0 ~pos:(placeholder_start + prefix_len) ~end_of_data)
+    | scanner_state -> (
+      (* We reached the end of the buffer: move the leftover data back to the
+         beginning of [buf] and refill the buffer *)
+      if leftover > 0 then
+        Bytes.blit ~src:buf ~dst:buf ~src_pos:placeholder_start ~dst_pos:0
+          ~len:leftover;
+      (* Reset [placeholder_start] to [0] since we moved back the leftover data
+         to the beginning of [buf] *)
+      let scanner_state : Scanner.state =
+        match scanner_state with
+        | Scan0
+        | Scan1
+        | Scan2 ->
+          scanner_state
+        | Scan_prefix _ -> Scan_prefix 0
+        | Scan_length (_, acc) -> Scan_length (0, acc)
+        | Scan_placeholder (_, len) -> Scan_placeholder (0, len)
+      in
+      match input buf leftover (buf_len - leftover) with
+      | 0 -> (
+        match scanner_state with
+        | Scan_placeholder _ ->
+          (* There might still be another placeholder after this invalid one
+             with a length that is too long *)
+          loop Scan0 ~pos:prefix_len ~end_of_data:leftover
+        | _ ->
+          (* Nothing more to read; [leftover] is definitely not the beginning of
+             a placeholder, send it and end the copy *)
+          Fiber.return false)
+      | n -> loop scanner_state ~pos:leftover ~end_of_data:(leftover + n))
+  in
+  match input buf 0 buf_len with
+  | 0 -> Fiber.return false
+  | n -> loop Scan0 ~pos:0 ~end_of_data:n
+
+let test_file ~src () =
+  let ic = Io.open_in src in
```

Comment:
This should be lazy IMO. It's cleaner to have all the side effects inside fiber.

Reply:
Consider this:

```ocaml
let test_src = test_file ~src () in
let* test_src = test_src ..
```
You're opening the channel after the first `let`, but clean up is reserved until the bind.

I would prefer if all the side effects occur after the bind.